libs/ledger/src/chain/main_chain.cpp

```
@@ -453,71 +557,126 @@ MainChain::Blocks MainChain::GetChainPreceding(BlockHash start, uint64_t limit)
     }
 
     // update the results
     result.push_back(std::move(block));
-
-    if (stop)
-    {
-      break;
-    }
   }
 
   return result;
 }
 
 /**
- * Walk the block history collecting blocks until either genesis or the block limit is reached.
- * Unlike in GetChainPreceding, positive value in limit indicates forward-travel.
+ * Walk the block history collecting blocks until either genesis or the block_number limit reached.
+ * limit specifies destination block_number.
+ * If current_hash is empty, travel starts from tip with zero limit, and from genesis otherwise.
  *
- * @param start The hash of the first block
- * @param limit The maximum number of blocks to be returned, negative for towards genesis, positive
- * for towards tip
+ * @param current_hash The hash of the first block
+ * @param limit
+ * @param direction -1 for towards genesis, +1, for towards tip
  * @return The array of blocks
  * @throws std::runtime_error if a block lookup occurs
  */
-MainChain::Blocks MainChain::TimeTravel(BlockHash start, int64_t limit) const
+MainChain::Travelogue MainChain::TimeTravel(BlockHash current_hash, uint64_t limit,
+                                            int64_t direction) const
 {
-  if (limit <= 0)
+  if (direction == 0)
   {
-    return GetChainPreceding(std::move(start), static_cast<uint64_t>(-limit));
+    return {Blocks{}, 0};
   }
 
-  auto const lim =
-      static_cast<std::size_t>(std::min(limit, static_cast<int64_t>(MainChain::UPPER_BOUND)));
-  MilliTimer myTimer("MainChain::ChainPreceding");
+  if (current_hash.empty() && direction > 0)
+  {
+    current_hash = chain::GENESIS_DIGEST;
+  }
+
+  IntBlockPtr block;
+  BlockHash   next_hash;
 
   FETCH_LOCK(lock_);
+  if (!current_hash.empty())
+  {
+    if (!LookupBlock(current_hash, block, direction > 0 ? &next_hash : nullptr))
+    {
+      FETCH_LOG_ERROR(LOGGING_NAME, "Block lookup failure for block: ", ToBase64(current_hash));
+      throw std::runtime_error("Failed to lookup block");
```

I think we should avoid throws in the code if possible. In this instance you can cause a node to throw (and probably crash with an RPC call containing an invalid hash

---

Do you suggest `GetChainPreceding` be fixed as well?

Might be deserving a separate jira for exception handling.

---

ah yes, it makes sense to fix GetChainPreceding in a seperate PR